### PR TITLE
Be more mindful about requests to Mesos and Marathon

### PR DESCRIPTION
**Pull request by ichernetsky**
- Do not fetch all Marathon apps just to check if a particular app exists.
- Do not fetch Mesos executor stats as often; cache them whenever possible instead.

### CODE REVIEW

File: autoscaler/api_client.py [hunk 1]
```diff
@@ -78,72 +77,59 @@ def authenticate(self):
...
+                    verify=False
+                )
+
+            self.log.debug("%s %s %s", method, path, response.status_code)
+
+            if response.status_code != 200:
+                if response.status_code == 401 and auth:
+                    self.log.info("Token expired. Re-authenticating to DC/OS")
+                    self.authenticate()
+                    return self.dcos_rest(method, path, data=data, auth=False)
```
Comment (timcharper) [inline autoscaler/api_client.py:119]: nice removal of unnecessary loop; this recursion should always terminate with at least one recursion?

File: autoscaler/agent_stats.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        Args:
+            task: marathon app task
+            agent: agent on which the task is run
+            n: statistics snapshot number
+        Returns:
+            statistics snapshot for the specific task running on the agent
+        """
+
+        agent_stats = self.stats.get(agent, [])
+        assert len(agent_stats) >= n
```
Comment (timcharper) [inline autoscaler/agent_stats.py]: will this throw an uncaught exception?

File: autoscaler/agent_stats.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        self.stats = {}
+
+    def get_task_stats(self, agent, task, n=0):
+        """ Get the performance metrics of the given task running on
+        the specified agent. If the n'th snapshot is cached, it is
+        returned, otherwise a request to the agent is made.
+        Args:
+            task: marathon app task
+            agent: agent on which the task is run
+            n: statistics snapshot number
```
Comment (timcharper) [inline autoscaler/agent_stats.py]: it's not clear what statistics snapshot number means here, can you help me understand?

File: autoscaler/api_client.py (hunk 1 above)
Comment (ichernetsky) [reply autoscaler/api_client.py:119]: Yes, you are correct.

File: autoscaler/agent_stats.py (hunk 2 above)
Comment (ichernetsky) [reply autoscaler/agent_stats.py]: It is not expected to be triggered. If it is, there is a bug in the code. All such exceptions are handled in `Autoscaler.run`

File: autoscaler/agent_stats.py (hunk 3 above)
Comment (ichernetsky) [reply autoscaler/agent_stats.py]: Some modes such as `ScaleByMemory` need only one snapshot of statistics for a task on an agent in order make a call when the corresponding app needs to be scale up or down or not. Whereas other modes such as `ScaleByCPU` need two such data points (snapshots) in order to figure out a task's CPU usage. Does it make it clear?

File: autoscaler/agent_stats.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
```
Comment (timcharper) [inline autoscaler/agent_stats.py]:
```suggestion
            n: statistics snapshot index
```

**Review by timcharper — APPROVED** [review]

**Pull request merged**